pandas/tools/merge.py

```diff
@@ -629,13 +808,136 @@ def get_result(self):
             concat_axis=0, copy=self.copy)
 
         typ = self.left._constructor
-        result = typ(result_data).__finalize__(self, method='ordered_merge')
+        result = typ(result_data).__finalize__(self, method=self._merge_type)
 
         self._maybe_add_join_keys(result, left_indexer, right_indexer)
 
         return result
 
 
+class _AsOfMerge(_OrderedMerge):
+    _merge_type = 'asof_merge'
+
+    def __init__(self, left, right, on=None, by=None, left_on=None,
+                 right_on=None, axis=1,
+                 suffixes=('_x', '_y'), copy=True,
+                 fill_method=None,
+                 how='asof', tolerance=None, strict=False):
+
+        self.by = by
+        self.tolerance = tolerance
+        self.strict = strict
+
+        _OrderedMerge.__init__(self, left, right, on=on, left_on=left_on,
+                               right_on=right_on, axis=axis,
+                               how=how, suffixes=suffixes,
+                               fill_method=fill_method)
+
+    def _validate_specification(self):
+        super(_AsOfMerge, self)._validate_specification()
+
+        # we only allow on to be a single item for on
+        if len(self.left_on) != 1:
+            raise MergeError("can only asof on a key for left")
+
+        if len(self.right_on) != 1:
+            raise MergeError("can only asof on a key for right")
+
+        # add by to our key-list so we can have it in the
+        # output as a key
+        if self.by is not None:
+            if not is_list_like(self.by):
+                self.by = [self.by]
+
+            self.left_on = self.by + list(self.left_on)
+            self.right_on = self.by + list(self.right_on)
+
+    @property
+    def _asof_key(self):
+        """ This is our asof key, the 'on' """
+        return self.left_on[-1]
+
+    def _get_merge_keys(self):
+
+        # note this function has side effects
+        (left_join_keys,
+         right_join_keys,
+         join_names) = super(_AsOfMerge, self)._get_merge_keys()
+
+        # validate index types are the same
+        for lk, rk in zip(left_join_keys, right_join_keys):
+            if not is_dtype_equal(lk.dtype, rk.dtype):
+                raise MergeError("incompatible merge keys, "
+                                 "must be the same type")
+
+        # no duplicates on rhs join keys
```

Comment:
its easy, my only question is this *too* magical. IOW would a user expect to get the dups? (e.g. it will work right now w/dups)